### ISSUE TITLE
HANG on aarch64, signal reproducibly not delivered to thread where it is unblocked

### ISSUE THREAD

**Issue by nsz-arm** [DynamoRIO/dynamorio#5927]
under drrun when a process blocks a signal, the signal sent to the process can remain pending forever even after some threads unblock it, because the dynamorio handler for the signal does not seem to deliver it to other threads.

the posix requirement is

"Signals generated for the process shall be delivered to exactly one of those threads within the process which is in a call to a [sigwait()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sigwait.html) function selecting that signal or has not blocked delivery of the signal. If there are no threads in a call to a [sigwait()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sigwait.html) function selecting that signal, and if all threads within the process block delivery of the signal, the signal shall remain pending on the process until a thread calls a [sigwait()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sigwait.html) function selecting that signal, a thread unblocks delivery of the signal, or the action associated with the signal is set to ignore the signal."

example:
```c
#include <pthread.h>
#include <signal.h>
#include <unistd.h>
#include <sys/wait.h>
#include <stdio.h>

#define die() do{ dprintf(1, "%s:%d failed\n", __FILE__, __LINE__); _exit(1); }while(0)

static void *start(void *a)
{
        sigset_t *ss = a;
        sigdelset(ss, SIGUSR1); /* ensure signal is delivered to this thread */
        if (pthread_sigmask(SIG_SETMASK, ss, 0)) die();
        pause();
        die();
        return 0;
}

int main()
{
        int r = fork();
        if (r == 0) {
                pthread_t t;
                sigset_t ss;
                sigfillset(&ss);
                sigdelset(&ss, SIGINT);  /* keep ctrl-c working */
                if (pthread_sigmask(SIG_SETMASK, &ss, 0)) die();
                if (pthread_create(&t, 0, start, &ss)) die();
                if (pthread_join(t,0)) die();
                die();
        }
        if (r < 0) die();
        if (usleep(42000)) die();
        if (kill(r, SIGUSR1)) die();
        int status;
        if (waitpid(r, &status, 0) != r) die();
        return 0;
}
```
running under strace:
```
$ strace -f ./kill2 2>&1 |grep -C5 USR1
[pid 1452646] clone(child_stack=0xffff8e737ae0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTIDstrace: Process 1452647 attached
, parent_tid=[1452647], tls=0xffff8e7388e0, child_tidptr=0xffff8e7382b0) = 1452647
[pid 1452647] set_robust_list(0xffff8e7382c0, 24 <unfinished ...>
[pid 1452646] futex(0xffff8e7382b0, FUTEX_WAIT, 1452647, NULL <unfinished ...>
[pid 1452647] <... set_robust_list resumed>) = 0
[pid 1452647] rt_sigprocmask(SIG_SETMASK, ~[INT USR1 RTMIN RT_1], NULL, 8) = 0
[pid 1452647] ppoll(NULL, 0, NULL, NULL, 0 <unfinished ...>
[pid 1452645] <... clock_nanosleep resumed>NULL) = 0
[pid 1452645] kill(1452646, SIGUSR1)    = 0
[pid 1452647] <... ppoll resumed>)      = ? ERESTARTNOHAND (To be restarted if no handler)
[pid 1452645] wait4(1452646,  <unfinished ...>
[pid 1452647] --- SIGUSR1 {si_signo=SIGUSR1, si_code=SI_USER, si_pid=1452645, si_uid=1000} ---
[pid 1452646] <... futex resumed>)      = ?
[pid 1452647] +++ killed by SIGUSR1 +++
[pid 1452646] +++ killed by SIGUSR1 +++
<... wait4 resumed>[{WIFSIGNALED(s) && WTERMSIG(s) == SIGUSR1}], 0, NULL) = 1452646
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=1452646, si_uid=1000, si_status=SIGUSR1, si_utime=0, si_stime=0} ---
exit_group(0)                           = ?
+++ exited with 0 +++
```
and with drrun:
```
$ strace -f ../DynamoRIO-AArch64-Linux-9.0.1/bin64/drrun ./kill2 2>&1 |grep -C5 USR1
rt_sigaction(SIGABRT, {sa_handler=0x7115f5a8, sa_mask=~[SEGV USR2], sa_flags=SA_RESTORER|SA_ONSTACK|SA_RESTART|SA_SIGINFO, sa_restorer=0x7115f58c}, NULL, 8) = 0
rt_sigaction(SIGBUS, NULL, {sa_handler=0x7115f5a8, sa_mask=~[KILL SEGV USR2 STOP], sa_flags=SA_RESTORER|SA_ONSTACK|SA_RESTART|SA_SIGINFO, sa_restorer=0x7115f58c}, 8) = 0
rt_sigaction(SIGBUS, {sa_handler=0x7115f5a8, sa_mask=~[SEGV USR2], sa_flags=SA_RESTORER|SA_ONSTACK|SA_RESTART|SA_SIGINFO, sa_restorer=0x7115f58c}, NULL, 8) = 0
rt_sigaction(SIGFPE, NULL, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGFPE, {sa_handler=0x7115f5a8, sa_mask=~[SEGV USR2], sa_flags=SA_RESTORER|SA_ONSTACK|SA_RESTART|SA_SIGINFO, sa_restorer=0x7115f58c}, NULL, 8) = 0
rt_sigaction(SIGUSR1, NULL, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGUSR1, {sa_handler=0x7115f5a8, sa_mask=~[SEGV USR2], sa_flags=SA_RESTORER|SA_ONSTACK|SA_RESTART|SA_SIGINFO, sa_restorer=0x7115f58c}, NULL, 8) = 0
rt_sigaction(SIGSEGV, NULL, {sa_handler=0x7115f5a8, sa_mask=~[KILL SEGV USR2 STOP], sa_flags=SA_RESTORER|SA_ONSTACK|SA_RESTART|SA_NODEFER|SA_SIGINFO, sa_restorer=0x7115f58c}, 8) = 0
rt_sigaction(SIGSEGV, {sa_handler=0x7115f5a8, sa_mask=~[SEGV USR2], sa_flags=SA_RESTORER|SA_ONSTACK|SA_RESTART|SA_NODEFER|SA_SIGINFO, sa_restorer=0x7115f58c}, NULL, 8) = 0
rt_sigaction(SIGUSR2, NULL, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGUSR2, {sa_handler=0x7115f5a8, sa_mask=~[SEGV USR2], sa_flags=SA_RESTORER|SA_ONSTACK|SA_RESTART|SA_NODEFER|SA_SIGINFO, sa_restorer=0x7115f58c}, NULL, 8) = 0
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
--
[pid 1452654] mprotect(0xfffd43cd0000, 4096, PROT_READ|PROT_WRITE) = 0
[pid 1452654] ppoll(NULL, 0, NULL, NULL, 0 <unfinished ...>
[pid 1452652] <... clock_nanosleep resumed>NULL) = 0
[pid 1452652] getpid()                  = 1452652
[pid 1452652] getpgid(0)                = 1452648
[pid 1452652] kill(1452653, SIGUSR1)    = 0
[pid 1452653] <... futex resumed>)      = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
[pid 1452653] --- SIGUSR1 {si_signo=SIGUSR1, si_code=SI_USER, si_pid=1452652, si_uid=1000} ---
[pid 1452653] mprotect(0xfffd43c77000, 4096, PROT_READ|PROT_WRITE <unfinished ...>
[pid 1452652] wait4(1452653,  <unfinished ...>
[pid 1452653] <... mprotect resumed>)   = 0
[pid 1452653] rt_sigreturn({mask=[]})   = 281472957235888
```
pid 1452653 gets stuck in the (interrupted) futex of pthread_join after the dynamorio SIGUSR1 handler returns, while pid 1452654 waits in pause and the signal never gets delivered there. the parent process 1452652 keeps waiting forever after it killed its child.

the correct behaviour is to deliver SIGUSR1 to the thread where it is unblocked and take the default action: terminate the child process.

**Comment by derekbruening** [comment]
This looks like a duplicate of #2311, right?  At least based on a quick reading it seems to be the same issue with DR not looking for a thread where it's unblocked for a signal sent to the whole process.
